// stablehorde/lib/models/StableHordeGenerations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StableHordeGenerations = void 0;
const StableHordeGeneration_1 = require("./StableHordeGeneration");
/**
 * @api public
 * @class StableHordeGenerations
 * @classdesc The list of generations returned by Stable Horde, as well as the
 *            workers that generated them.
 */
class StableHordeGenerations {
    constructor(generations) {
        this.generations = generations;
    }
    toJSON() {
        return {
            generations: this.generations.map((g) => g.toJSON()),
        };
    }
    static fromJSON(data) {
        return new StableHordeGenerations(data['generations'].map((g) => StableHordeGeneration_1.StableHordeGeneration.fromJSON(g)));
    }
    /**
     * The generated images as WEBP binary buffers.
     *
     * @returns {Buffer[]} An array of buffers containing the WEBP image data.
     */
    buffers() {
        return this.generations.map((g) => g.buffer());
    }
    /**
     * The generated images as base64-encoded strings.
     *
     * @returns {string[]} An array of base64-encoded WEBP images.
     */
    base64() {
        return this.generations.map((g) => g.base64());
    }
}
exports.StableHordeGenerations = StableHordeGenerations;
