// stablehorde/lib/models/StableHordeRequestStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StableHordeRequestStatus = void 0;
/**
 * @api public
 * @class StableHordeRequestStatus
 * @classdesc Status information about an image generation request.
 */
class StableHordeRequestStatus {
    constructor(done, waiting, processing, finished, queuePosition, waitTime) {
        this.done = done;
        this.waiting = waiting;
        this.processing = processing;
        this.finished = finished;
        this.queuePosition = queuePosition;
        this.waitTime = waitTime;
    }
    toJSON() {
        return {
            done: this.done,
            waiting: this.waiting,
            processing: this.processing,
            finished: this.finished,
            queue_position: this.queuePosition,
            wait_time: this.waitTime,
        };
    }
    static fromJSON(data) {
        return new StableHordeRequestStatus(data['done'], data['waiting'], data['processing'], data['finished'], data['queue_position'], data['wait_time']);
    }
}
exports.StableHordeRequestStatus = StableHordeRequestStatus;
